### DISCUSSION TITLE
h3 server looses cookies on page-load/page-reload

### DISCUSSION THREAD

**Discussion by folamy** [nuxt/nuxt#17953] [Questions] [answered]
### Environment

- Operating System: Linux OS/Fedora
- Node Version: v16.11.0
- Nuxt Version: 3.0.0-27385241.dcc0c9
- Package Manager: yarn@1.22.17
- Bundler: Vite

### Reproduction

nil

### Describe the bug

Having a page where I need to get user's session/cookie data from the server/backend:
Firstly, on SSR, there is no **Cookie** on useNuxtApp().ssrContext.req.headers! (This would have sufficed for my use case)

secondly, I tried a workaround option like this

```js
// server/api/student/reqister
import type { IncomingMessage, ServerResponse } from 'http'
import { setCookie } from 'h3

export default async (req: IncomingMessage, res: ServerResponse) => {
  const body = await useBody(req)
  // do a db operation that returns an object containing registered student
  const { registeredStudent } = await AddStudent(body)
   setCookie(res, 'registeredStudent', JSON.stringify(registeredStudent))
  return registeredStudent
}

// server/api/student/get-data
import type { IncomingMessage, ServerResponse } from 'http'
import { useCookies } from 'h3

export default async (req: IncomingMessage, res: ServerResponse) => {
  return useCookies(req)
}

// pages/student/confirm-user-registration
<script setup lang="ts">
const { data } =  await useAsyncData('get-user-data', () => $fetch('/api/student/get-data'))
console.log(data.value) 
// this works on router navigation within the browser
// But it doesn't work on page reload 
</script>
```
I need to get the data before Vue renders the page

**Comment by danielroe** [comment]
Try using `useRequestHeaders()`. This should explain why: http://v3.nuxtjs.org/docs/usage/data-fetching#isomorphic-fetch

Note that if you also want to _set_ a cookie on SSR you'll need to do so yourself within asyncData.

**Comment by folamy** [comment]
> Try using `useRequestHeaders()`. This should explain why: http://v3.nuxtjs.org/docs/usage/data-fetching#isomorphic-fetch
> 
> Note that if you also want to _set_ a cookie on SSR you'll need to do so yourself within asyncData.

are you suggesting that I should do this? 

```js
const { data } = useFetch(/api/student/get-data', {
  headers: useRequestHeaders(['cookie'])
})
```

**Reply by danielroe** [reply]
Give it a try. If it's not working, you can create a reproduction and I or someone else may be able to help.

**Reply by folamy** [reply]
> Give it a try. If it's not working, you can create a reproduction and I or someone else may be able to help.

@danielroe, I tried on my local machine, but it didn't work! So, I created [This on stackblitz](https://stackblitz.com/edit/github-ui4bax-7zixtm)
https://stackblitz.com/edit/github-ui4bax-7zixtm

Unfortunately, the useReqestHeadrs() doesn't work on stackblitz!

**Comment by Diizzayy** [comment, marked as answer]
@folamy I had a look at your issue, it's actually very simple to understand and rectify.

Firstly, when using setCookie within API Routes you must specify the `cookie path` or else it would default to `/api` (_the problem you faced here_, hence the cookie will only be available on the api route).

This can be addressed as seen below, making this change will allow your cookie to be set and utilized as expected when using API Routes.
```ts
  setCookie(res, "registered", JSON.stringify(registered), {
    path: "/",
  });
```

> on SSR, there is no Cookie on useNuxtApp().ssrContext.req.headers! (This would have sufficed for my use case)
- Now as it pertains to this, on SSR, the initial load as you know happens on the server and by default it has zero knowledge of what cookies the client (your browser) may have. `useRequestHeaders()` is indeed the way to get around this by proxying the pertinent headers from your client (`cookie` in this case) to your server as seen below.
```js
<script setup lang="ts">
const { data } = await useFetch("/api/get-registered", {
  headers: useRequestHeaders(["cookie"]),
});
console.log(data.value);
</script>
```
Also, an important note, allow both `useFetch` and `useRequestHeaders` to **be auto-imported by nuxt** to avoid running into weird issues that you may encounter. (for example `import { useRequestHeaders } from "nuxt3";` as seen in your stackblitz demo will crash nuxt)

**Reply by viii47** [reply]
Nice writeup @Diizzayy. I'd add that if you do _not_ set a `maxAge: Number (in seconds)` property the Cookie will only be available on the current session and deleted after closing the browser/opening a new tab.

**Reply by folamy** [reply]
Wow! @Diizzayy, you rock this. Thanks for detail explanation.

**Reply by ghost** [reply]
The problem is that a useFetch call on the browser side is required for Nuxt/Nitro to accept to deliver the cookie to the browser. This means that on each SSR visit, a "new" client cannot be given a cookie, so it is absolutely necessary to create a no-ssr call for the client to get its cookie. This implies that for each new SSR visitor, a cookie is issued and information is created in Redis/database that will never be used because the browser will never receive it

No viable alternative to that? @Diizzayy
